File: src/dantofa/clients/digitalocean/clusters.py
"""Adapter over the pydo SDK for DigitalOcean Kubernetes (DOKS) clusters.

This lives in the clients layer because pydo is an untyped third-party SDK, so
values crossing this boundary are ``Any`` (see the basedpyright execution
environment for ``src/dantofa/clients`` in pyproject). The SDK is imported lazily
so importing the CLI does not pull in pydo's heavy azure-core transport until a
cluster command actually runs.
"""

from __future__ import annotations

import os
from collections.abc import Mapping
from typing import Any, final

from dantofa.clients.digitalocean.errors import (
    DigitalOceanApiError,
    MissingCredentialsError,
)

TOKEN_ENV = "DIGITALOCEAN_ACCESS_TOKEN"

# DigitalOcean caps per_page at 200; request the max to minimise round-trips.
_PER_PAGE = 200

# pydo's generated get_kubeconfig parses the response as JSON, which fails on the
# YAML kubeconfig, so that one endpoint is called directly against the REST API.
_API_BASE = "https://api.digitalocean.com"


def _resolve_token(token: str | None) -> str:
    resolved = token or os.environ.get(TOKEN_ENV)
    if not resolved:
        raise MissingCredentialsError(
            f"No DigitalOcean API token. Pass --token or set ${TOKEN_ENV}.",
        )
    return resolved


def _api_error(exc: Any) -> DigitalOceanApiError:
    """Wrap an azure-core error, preserving DigitalOcean's raw JSON error body."""
    response = getattr(exc, "response", None)
    if response is not None:
        try:
            return DigitalOceanApiError(response.json())
        except Exception:  # noqa: BLE001 - body absent/non-JSON; fall back to text
            return DigitalOceanApiError(str(exc))
    return DigitalOceanApiError(str(exc))


@final
class ClusterClient:
    """Semantic wrapper over pydo's ``kubernetes`` operations.

    Returns plain dicts/lists; raises :class:`DigitalOceanApiError` on API
    failures so callers stay free of pydo/azure-core error types.
    """

    def __init__(self, token: str | None = None) -> None:
        from pydo import Client  # noqa: PLC0415 — lazy: avoid eager azure-core import

        self._token = _resolve_token(token)
        self._client = Client(token=self._token)

    def list(self) -> list[dict[str, Any]]:
        """Return every cluster, following pagination links."""
        from azure.core.exceptions import HttpResponseError  # noqa: PLC0415

        clusters: list[dict[str, Any]] = []
        page = 1
        try:
            while True:
                resp = self._client.kubernetes.list_clusters(
                    per_page=_PER_PAGE,
                    page=page,
                )
                batch = list(resp.get("kubernetes_clusters") or [])
                clusters.extend(batch)
                links: dict[str, Any] = resp.get("links") or {}
                pages: dict[str, Any] = links.get("pages") or {}
                if not batch or not pages.get("next"):
                    break
                page += 1
        except HttpResponseError as exc:
            raise _api_error(exc) from exc
        return clusters

    def create(self, body: Mapping[str, Any]) -> dict[str, Any]:
        from azure.core.exceptions import HttpResponseError  # noqa: PLC0415

        try:
            return dict(self._client.kubernetes.create_cluster(body=dict(body)))
        except HttpResponseError as exc:
            raise _api_error(exc) from exc

    def update(self, cluster_id: str, body: Mapping[str, Any]) -> dict[str, Any]:
        from azure.core.exceptions import HttpResponseError  # noqa: PLC0415

        try:
            return dict(
                self._client.kubernetes.update_cluster(
                    cluster_id=cluster_id,
                    body=dict(body),
                ),
            )
        except HttpResponseError as exc:
            raise _api_error(exc) from exc

    def delete(self, cluster_id: str) -> None:
        from azure.core.exceptions import HttpResponseError  # noqa: PLC0415

        try:
            _ = self._client.kubernetes.delete_cluster(cluster_id=cluster_id)
        except HttpResponseError as exc:
            raise _api_error(exc) from exc

    def get(self, cluster_id: str) -> dict[str, Any]:
        """Return a single cluster (includes ``status.state``)."""
        from azure.core.exceptions import HttpResponseError  # noqa: PLC0415

        try:
            resp = self._client.kubernetes.get_cluster(cluster_id=cluster_id)
        except HttpResponseError as exc:
            raise _api_error(exc) from exc
        return dict(resp.get("kubernetes_cluster") or {})

    def get_kubeconfig(self, cluster_id: str) -> str:
        """Return the cluster's kubeconfig YAML (fetched directly; see _API_BASE)."""
        import httpx  # noqa: PLC0415 — lazy import to keep CLI startup fast

        # SKY-D216 false positive: the host is the fixed _API_BASE constant; only
        # the path carries cluster_id (a DigitalOcean-issued id), so there is no
        # attacker-controlled URL/host here.
        response = httpx.get(  # skylos: ignore[SKY-D216]
            f"{_API_BASE}/v2/kubernetes/clusters/{cluster_id}/kubeconfig",
            headers={"Authorization": f"Bearer {self._token}"},
            timeout=30.0,
        )
        if response.is_success:
            return response.text
        try:
            payload: object = response.json()
        except Exception:  # noqa: BLE001 - non-JSON error body; fall back to text
            payload = response.text
        raise DigitalOceanApiError(payload)

File: internal/clients/digitalocean/clusters.go
package digitalocean

import (
	"context"
	"fmt"
	"os"

	"github.com/digitalocean/godo"

	core "github.com/dantofa/platform/internal/core/digitalocean"
)

const tokenEnv = "DIGITALOCEAN_ACCESS_TOKEN"

// perPage requests DigitalOcean's max page size to minimise round-trips.
const perPage = 200

// ClusterClient is a semantic wrapper over godo's Kubernetes operations. It
// returns core domain types and translates godo errors into APIError so callers
// stay free of the SDK's error types.
type ClusterClient struct {
	godo *godo.Client
}

// resolveDOToken returns the DO API token from the argument or
// $DIGITALOCEAN_ACCESS_TOKEN, or "" if neither is set.
func resolveDOToken(token string) string {
	if token == "" {
		return os.Getenv(tokenEnv)
	}
	return token
}

// NewClusterClient builds a cluster client, reading the token from the argument
// or $DIGITALOCEAN_ACCESS_TOKEN.
func NewClusterClient(token string) (*ClusterClient, error) {
	token = resolveDOToken(token)
	if token == "" {
		return nil, MissingCredentials(
			fmt.Sprintf("pass --token or set $%s.", tokenEnv),
		)
	}
	return &ClusterClient{godo: godo.NewFromToken(token)}, nil
}

// List returns every cluster, following pagination.
func (c *ClusterClient) List(ctx context.Context) ([]core.Cluster, error) {
	opts := &godo.ListOptions{Page: 1, PerPage: perPage}
	var clusters []core.Cluster
	for {
		page, resp, err := c.godo.Kubernetes.List(ctx, opts)
		if err != nil {
			return nil, apiError(err)
		}
		for _, cl := range page {
			clusters = append(clusters, toCoreCluster(cl))
		}
		if resp == nil || resp.Links == nil || resp.Links.IsLastPage() {
			break
		}
		next, err := resp.Links.CurrentPage()
		if err != nil {
			return nil, err
		}
		opts.Page = next + 1
	}
	return clusters, nil
}

// Create creates a cluster from the neutral spec.
func (c *ClusterClient) Create(ctx context.Context, spec core.CreateSpec) (core.Cluster, error) {
	pools := make([]*godo.KubernetesNodePoolCreateRequest, 0, len(spec.NodePools))
	for _, p := range spec.NodePools {
		pools = append(pools, &godo.KubernetesNodePoolCreateRequest{
			Name: p.Name, Size: p.Size, Count: p.Count,
			AutoScale: p.AutoScale, MinNodes: p.MinNodes, MaxNodes: p.MaxNodes,
		})
	}
	req := &godo.KubernetesClusterCreateRequest{
		Name:         spec.Name,
		RegionSlug:   spec.Region,
		VersionSlug:  spec.Version,
		Tags:         spec.Tags,
		NodePools:    pools,
		AutoUpgrade:  spec.AutoUpgrade,
		SurgeUpgrade: spec.SurgeUpgrade,
	}
	if spec.HA {
		ha := true
		req.HA = &ha
	}
	cl, _, err := c.godo.Kubernetes.Create(ctx, req)
	if err != nil {
		return core.Cluster{}, apiError(err)
	}
	return toCoreCluster(cl), nil
}

// Update updates a cluster's mutable fields.
func (c *ClusterClient) Update(ctx context.Context, id string, spec core.UpdateSpec) (core.Cluster, error) {
	autoUpgrade := spec.AutoUpgrade
	req := &godo.KubernetesClusterUpdateRequest{
		Name:         spec.Name,
		AutoUpgrade:  &autoUpgrade,
		SurgeUpgrade: spec.SurgeUpgrade,
	}
	if spec.Tags != nil {
		req.Tags = *spec.Tags
	}
	if spec.HA {
		ha := true
		req.HA = &ha
	}
	cl, _, err := c.godo.Kubernetes.Update(ctx, id, req)
	if err != nil {
		return core.Cluster{}, apiError(err)
	}
	return toCoreCluster(cl), nil
}

// Delete deletes a cluster by id.
func (c *ClusterClient) Delete(ctx context.Context, id string) error {
	_, err := c.godo.Kubernetes.Delete(ctx, id)
	if err != nil {
		return apiError(err)
	}
	return nil
}

// Get returns a single cluster (including status.state).
func (c *ClusterClient) Get(ctx context.Context, id string) (core.Cluster, error) {
	cl, _, err := c.godo.Kubernetes.Get(ctx, id)
	if err != nil {
		return core.Cluster{}, apiError(err)
	}
	return toCoreCluster(cl), nil
}

// GetKubeconfig returns the cluster's kubeconfig YAML.
func (c *ClusterClient) GetKubeconfig(ctx context.Context, id string) (string, error) {
	cfg, _, err := c.godo.Kubernetes.GetKubeConfig(ctx, id, nil)
	if err != nil {
		return "", apiError(err)
	}
	return string(cfg.KubeconfigYAML), nil
}

func toCoreCluster(cl *godo.KubernetesCluster) core.Cluster {
	if cl == nil {
		return core.Cluster{}
	}
	out := core.Cluster{
		ID:       cl.ID,
		Name:     cl.Name,
		Region:   cl.RegionSlug,
		Version:  cl.VersionSlug,
		Endpoint: cl.Endpoint,
		Tags:     cl.Tags,
	}
	if cl.Status != nil {
		out.State = string(cl.Status.State)
	}
	if !cl.CreatedAt.IsZero() {
		out.CreatedAt = cl.CreatedAt.Format("2006-01-02T15:04:05Z07:00")
	}
	for _, p := range cl.NodePools {
		if p == nil {
			continue
		}
		out.NodePools = append(out.NodePools, core.NodePool{
			ID: p.ID, Name: p.Name, Size: p.Size, Count: p.Count,
			AutoScale: p.AutoScale, MinNodes: p.MinNodes, MaxNodes: p.MaxNodes,
		})
	}
	return out
}
